File: Leet 55/solution.py
import functools

class Solution:
    def canJump(self, nums: list[int]) -> bool:

        @functools.cache
        def jump_to(i: int = 0) -> bool:
            if i >= len(nums) - 1:
                return True

            amount = i + nums[i]
            if amount > len(nums) - 1:
                amount = len(nums) - 1
             
            for j in range(i + 1, amount + 1):
                if jump_to(j):
                    return True
                
            return False
    
        return jump_to()
    

def main() -> None:
    print("55. Jump Game")
    print(Solution().canJump([2,3,1,1,4]))


if __name__ == "__main__":
    main()

File: Leet 55/solution.cs
public class Solution
{
    public bool CanJump(int[] nums)
    {
        int last = nums.Length - 1;
        for (int i = nums.Length - 1; i >= 0; i--)
        {
            if (i + nums[i] >= last)
                last = i;
        }
        return last == 0;
    }
}

class Program
{
    private static void Main()
    {
        Console.WriteLine("55. Jump Game");

        int[] nums = [2, 3, 1, 1, 4];
        
        Solution solution = new();
        string result = solution.CanJump(nums) ? "Pass" : "Fail";
        Console.WriteLine($"Can Jump should pass ? {result}");
    }
}